### ISSUE TITLE
Hitting LocalPRG assert statement

### ISSUE THREAD

**Issue by mbhall88** [rmcolq/pandora#126]
I have a PRG which is triggering this `assert` statement https://github.com/rmcolq/pandora/blob/7adf63ce60d28a000f7b6c850f4a5ebbbf2dd031/src/localPRG.cpp#L712-L713

I can't seem to figure out why/how though. I think the gene which triggers the `assert` statement is `GC00000119_18`. Although it is hard to tell as the assert error is raised after quite a bit more code has run.

Also not sure how to get around it.  Is there a way we can guard against this happening?

## Error message

```
pandora: /nfs/research1/zi/mbhall/Software/pandora/src/localPRG.cpp:714: std::vector<unsigned int> get_covgs_along_localnode_path(PanNodePtr, const std::vector<std::shared_ptr<LocalNode> >&, const std::vector<std::shared_ptr<KmerNode> >&, const uint32_t&): Assertion `localnode_path[k]->pos.start <= interval.start and localnode_path[k]->pos.get_end() >= interval.get_end()' failed.
```
*Note: the line number stated in this error message differs slightly from the lines I have linked to above as the commit I am using has an extra logging statement.*

## Steps to reproduce
On ebi-cli:
```sh
ref=/hps/nobackup/research/zi/mbhall/pandora_simulations/analysis/1/50/combined_random_paths_mutated_1.fasta
genome_size=$(grep -v '>'  "$ref" | wc | awk '{print $3-$1-10}')
read_file=/hps/nobackup/research/zi/mbhall/pandora_simulations/analysis/1/50/imperfect/reads.simulated.fa
prg_file=/hps/nobackup/research/zi/mbhall/pandora_simulations/data/prgs/max_nesting_lvl_1/combined.prg.fa
outdir=tmp/
coverage=100
denovo_kmer_size=13
log=map.log
        
/nfs/research1/zi/mbhall/Software/bin/pandora map --prg_file "$prg_file" \
        --read_file "$read_file" \
        --outdir "$outdir" \
        --output_kg \
        --output_covgs \
        --max_covg "$coverage" \
        --output_vcf \
        --genotype \
        --genome_size "$genome_size" \
        --discover \
        --denovo_kmer_size "$denovo_kmer_size" \
        --log_level debug &> "$log"
```

**Comment by leoisl** [comment]
hello,

I looked a little bit at this, but I don't really know pandora code - I still have to sit and go through it (first time looking at it!). With debug, it might have some info that you might understand and interest you. Sorry if you already saw these debug values, as this info will then be no news for you. I don't know the code, so can't help you much, but in some days I will start studying it, I'd appreciate some discussions on it! I though also screenshots from the debug IDE could be nice to see the context, etc (but maybe text debug would be better)...

The gene which triggers the assert seems to be GC00000596_2 (prg_id=2):
![1](https://user-images.githubusercontent.com/8397967/57414312-812f6480-71ef-11e9-8cf6-d778defcfe0f.png)

The interval (from a kmernode) where it bugs has start 11988 and length 15:
![2](https://user-images.githubusercontent.com/8397967/57414326-8be9f980-71ef-11e9-81f0-880fc424e7b9.png)

`localnode_path[k]->pos` has start 10663 and length 1334.
Therefore, the assertion fails on localPRG.cpp:713, since 10663+1334 is not >= 11988+15, ie the interval is not contained in `localnode_path[k]->pos`:
![3](https://user-images.githubusercontent.com/8397967/57414340-94dacb00-71ef-11e9-9e70-6002bb6ffd49.png)

These are very basic infos, maybe you already know. I dug deeper in other parts also, but I should understand better these parts before saying anything concrete...

cheers

**Comment by rmcolq** [comment]
Thanks, this is helpful. Is the kmernode path along the local node path? They should be equivalent and not deviate from each other.



Sent from my Samsung Galaxy smartphone.


-------- Original message --------

**Comment by leoisl** [comment]
Hello there!

I think the kmernode path is not along the localnode path. This is what I mean (I have also several doubts in my explanation, so not sure of anything!):

The bug is on processing the first kmernode of the kmernode path (is kmernode path a path of minimizers kmers from a read?). This kmernode itself has a path of intervals (are these the intervals of the PRG that compose the kmer?). The intervals (described as (start, length)) of this kmernode are: (0,0) and (11988, 15).

Is localnode path a path of nodes of the PanRG where each kmer of the kmernode path is in? The interval of the first localnode in the localnode path is (0,0), matching the first interval of the kmernode. The interval of the second local node is (10663, 1334). The interval of the third local node is (12000, 1334). The second interval of the kmernode ((11988, 15)) is partly present in the second local node (from 11988 till 11997) and in the third local node (from 12000 till 12003). There is also a part of this kmernode that is not present in either intervals. I guess this was not supposed to happen, there should be a local node containing  this kmernode?

**Comment by rmcolq** [comment]
Great! Michael, us this function being called as part of usual pandora map or compare (in which case there may be a bug with the code identifying the local node path from the kmernode max path) or elsewhere?
Rachel



Sent from my Samsung Galaxy smartphone.


-------- Original message --------

**Comment by leoisl** [comment]
hello,

I can try to debug further, but some high-level explanations would be really helpful. I have some guesses, other stuff I have no idea... Could you tell me:

1. is kmernode path a path of minimizers kmers from a read?
2. A kmernode itself has a path of intervals. Are these the intervals of the PRG that compose the kmer?
3. What a  localnode path represents?

**Comment by rmcolq** [comment]
2. Is correct
3. A prg is a directed acyclic sequence graph with a single source and sink (a null node if necessary). A local node path is a single path along nodes through this graph usually all the way from source to sink.
1. Kmernode path should be the ordered list of kmer nodes along a single local node path. The max path algorithm finds the path through the prg first as a kmernode path which has to be translated back to the local node path.

Rachel



Sent from my Samsung Galaxy smartphone.


-------- Original message --------

**Comment by leoisl** [comment]
Great, that clear things up! Thanks a lot Rachel!

**Comment by leoisl** [comment]
hello (even though you are just behind me :p)

I guess the assertion fails because when translating the kmernode path to the localnode path, it is enough that local nodes just overlaps the kmer (more specifically, overlaps any interval of the kmer): https://github.com/rmcolq/pandora/blob/master/src/localPRG.cpp#L83 . When collecting the coverage, it is required that the local node contains the kmer (stronger condition): https://github.com/rmcolq/pandora/blob/master/src/localPRG.cpp#L712-L713 . Some local nodes overlaps with the kmer without containing it, and this is when the assertion fails.

However, I don't know what would be the correct fix. I am not sure if the first condition should be changed to allow only local nodes containing the k-mers (not only overlapping) or the second should be changed to allow also overlapping (but not only contained) local nodes. My blunt guess is that the assertion should check only for overlapping. Do you know what to do?

thanks for all the help!

**Comment by rmcolq** [comment]
I still think we need to know what Michael is doing at the time this function is being called (and I guess that will have to wait until he has finished travelling). 

The function with the assert requires that the kmernode path and localnode path are equivalent. I don't want to change that. And the input being given is not equivalent.

When the kmernode path is translated to a localnode path, as you have gathered, we infer the localnodes overlapped by each kmernode, and piece together overlapping lists of local nodes. Given that a kmernode is a path of length k in the PRG, I don't think it should be possible for it to overlap another localnode which is incorrect.

Something weird is happening

**Comment by leoisl** [comment]
You are right, the assert should not be changed. We talk to Michael when he comes!

+1 for sth weird is happening

Thanks thanks!!

**Comment by rmcolq** [comment]
Maybe look at `subpath', which is called e.g. here https://github.com/rmcolq/pandora/blob/7adf63ce60d28a000f7b6c850f4a5ebbbf2dd031/src/localPRG.cpp#L359 and defined here https://github.com/rmcolq/pandora/blob/7adf63ce60d28a000f7b6c850f4a5ebbbf2dd031/src/prg/path.cpp#L56

**Comment by leoisl** [comment]
Hello @mbhall88 ,

talking to Rachel, we think it could be a bug on the graph sketching, so could you send us your pandora index command also?

**Comment by leoisl** [comment]
TL;DR of all these exchanges:

A kmer (described by path: (0,0), (11988,15)) was sketched in the graph during pandora index, but its second interval does not exist in any local node (overlaps partially two local nodes: (10663, 1334) and  (12000, 1334)). kmers are sketched from the graph, can traverse several local nodes, but their intervals should be contained in a specific local node, so this should not happen.

I will debug pandora index step when getting the pandora index command from Michael to understand how this kmer was sketched.

**Comment by mbhall88** [comment]
I put the exact steps to reproduce in the [first comment](https://github.com/rmcolq/pandora/issues/126#issue-441042060)...  
As for the index command, I just ran index with default settings.

**Comment by mbhall88** [comment]
Very sorry. But it turns out that I had somehow duplicated the culprit gene in the PRG. Removing the duplicate lead to `map` running successfully.

**Comment by rmcolq** [comment]
Great! There needs to be something to catch this user error then somewhere - can you find a suitable place for an assert?

**Comment by mbhall88** [comment]
You mean if a gene is duplicated?

**Comment by rmcolq** [comment]
Yes. Was it duplicate of gene name that was the problem? I can't understand why right now but there is some warning when indexing if that is the case.

**Comment by mbhall88** [comment]
It was a duplicate of gene name and content. I didn't see a warning in the `index` log about the duplicate. But it makes the most sense to have a warning/error in there if you notice duplicate sequence names.

**Comment by leoisl** [comment]
Does this still compose a bug? What I mean is that if we had the exactly same PRG, but with a different name, the tool would work fine.

In some places, it seems that the name is important, e.g. when saving the gfa in kmer_prgs:
```
ls kmer_prgs/*
GC00000085_5.k15.w14.gfa  GC00002569.k15.w14.gfa  GC00007068_2.k15.w14.gfa
```

I am wondering if a simple solution like concatenating the prg id to the name would work, since it would make all names unique...

I can debug this, and provide a solution (i.e. make it work even with equal names) if you wish. This should be a nice exercise to understand better index and map...

**Comment by mbhall88** [comment]
My feeling is that if we see two sequences with the same name we should raise an error. A lot of other bioinf tools also exhibit this behaviour. I think it also saves bigger unexplained issues (like this one) arising further down the analysis road.

**Comment by iqbal-lab** [comment]
Error and die? or error and carry on but ignore the second one? arguments for both, but i sort of prefer the former

**Comment by rmcolq** [comment]
I thought I was doing the second of these. The index step should be issuing a warning that only the last sequence is kept. I haven't looked to see exactly what about having two things with the same name has caused the problem.

**Comment by leoisl** [comment]
I totally agree with Michael on this: https://github.com/rmcolq/pandora/issues/126#issuecomment-492161063 , so +1 for error and die.

**Comment by mbhall88** [comment]
> I thought I was doing the second of these. The index step should be issuing a warning that only the last sequence is kept. I haven't looked to see exactly what about having two things with the same name has caused the problem.

I can't see anywhere in the `index` that checks this...? Maybe you could point me to it (when you get time)?

**Comment by rmcolq** [comment]
My mistake, I don't. I do check that sample ids are not the same as each other in compare (`load_read_index`)...
